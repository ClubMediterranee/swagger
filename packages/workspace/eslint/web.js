module.exports = {
  root: true,
  env: {
    browser: true,
    es2021: true
  },
  extends: [
    "react-app",
    "react-app/jest",
    "plugin:jsx-a11y/strict",
    "plugin:prettier/recommended",
    "plugin:cypress/recommended"
  ],
  plugins: ["prettier", "react-hooks"],
  rules: {
    "@typescript-eslint/no-unused-vars": "error",
    "prettier/prettier": "error",
    "react-hooks/rules-of-hooks": "error", // Checks rules of Hooks
    "react-hooks/exhaustive-deps": "warn", // Checks effect dependencies
    "cypress/assertion-before-screenshot": "warn",
    "cypress/no-force": "warn",
    "cypress/no-pause": "error",
    "testing-library/no-unnecessary-act": "warn",
    "testing-library/prefer-screen-queries": "warn",
    "testing-library/prefer-presence-queries": "warn",
    "testing-library/no-wait-for-multiple-assertions": "warn",
    "testing-library/no-node-access": ["warn", { allowContainerFirstChild: true }],
    "testing-library/no-render-in-setup": "warn"
  }
};
