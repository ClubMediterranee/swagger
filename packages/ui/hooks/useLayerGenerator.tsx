import { useSpring, useSprings } from '@react-spring/web';

import { ComponentProps, useEffect, useRef, useState } from 'react';

import { Layer } from '../molecules/Layer';

function getWindowDimensions() {
  if (typeof window === 'undefined') {
    return {
      width: 0,
      height: 0,
    };
  }

  const { innerWidth: width, innerHeight: height } = window;

  return {
    width,
    height,
  };
}

function useWindowDimensions() {
  const [windowDimensions, setWindowDimensions] = useState(getWindowDimensions());

  useEffect(() => {
    function handleResize() {
      setWindowDimensions(getWindowDimensions());
    }

    window.addEventListener('resize', handleResize, { passive: true });

    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return windowDimensions;
}

const animatedSpringConfig = {
  mass: 1.5,
  tension: 180,
  friction: 30,
};

const initialPaintStyles = {
  insetInlineStart: 0,
  top: 0,
  transformOrigin: 'top left',
};

const initialLayerStyles = {
  transform: 'scale(1, 1)',
  willChange: 'transform, opacity, inset',
  pointerEvents: 'none',
  opacity: 0,
  zIndex: -1,
  ...initialPaintStyles,
  x: '-100%',
};

const initialTriggerStyles = {
  transform: 'scale(1, 1)',
  pointerEvents: 'auto',
  willChange: 'transform, opacity, inset, filter',
  opacity: 1,
  position: 'relative',
  filter: 'blur(0px)',
  zIndex: 0,
  ...initialPaintStyles,
};

let instances = 0;

type Content = Pick<ComponentProps<typeof Layer>, 'content'>['content'];

type LayerContent = { content: Content | ((closeLayer: () => void) => Content) };

type LayerProps = Pick<
  ComponentProps<typeof Layer>,
  'title' | 'closeLabel' | 'leftButtonProps' | 'rightButtonProps'
> &
  LayerContent & {
    placeholder?: Content;
  };

const useLayerGenerator = (layerProps: LayerProps, nbTriggers = 1, exitCallback?: () => void) => {
  const lastKnownTriggerBounds = useRef({ top: 0, left: 0, width: 0, height: 0 });
  const lastKnownRef = useRef(null);
  const [isOpen, setIsOpen] = useState(false);
  const hasBeenOpened = useRef(false);
  const { width: windowWidth, height: windowHeight } = useWindowDimensions();

  const [layerStyles, layerStylesAPI] = useSpring(() => ({
    from: initialLayerStyles,
    immediate: true,
  }));
  const [triggerStyles, triggerStylesAPI] = useSprings(nbTriggers, () => ({
    from: initialTriggerStyles,
    immediate: true,
  }));

  const openLayer = (rect: DOMRect | 'outside', index = 0) => {
    // TODO: fix it to use direction context
    const isRTL = document.documentElement.dir === 'rtl';
    const bounds =
      rect === 'outside'
        ? {
            top: 0,
            left: -1 * windowWidth,
            height: windowHeight,
            width: windowWidth,
          }
        : rect;

    hasBeenOpened.current = true;
    setIsOpen(true);

    lastKnownTriggerBounds.current = bounds;

    layerStylesAPI.start({
      from: {
        insetInlineStart: isRTL ? -1 * bounds.left : bounds.left,
        top: bounds.top,
        transform: `scale(${bounds.width / windowWidth}, ${bounds.height / windowHeight})`,
        pointerEvents: 'none',
        opacity: 0.2,
        transformOrigin: 'top left',
        zIndex: 1 + instances,
        x: '0%',
      },
      to: {
        insetInlineStart: 0,
        top: 0,
        transform: 'scale(1, 1)',
        pointerEvents: 'auto',
        opacity: 1,
        transformOrigin: 'top left',
        zIndex: 1 + instances,
      },
      config: animatedSpringConfig,
    });

    triggerStylesAPI.start((i) => {
      if (i === index) {
        return {
          from: { ...initialTriggerStyles, zIndex: 10 },
          to: {
            insetInlineStart: isRTL ? bounds.left : -1 * bounds.left,
            top: -1 * bounds.top,
            transform: `scale(${windowWidth / bounds.width}, ${windowWidth / bounds.width})`,
            pointerEvents: 'none',
            opacity: 0,
            filter: 'blur(10px)',
            transformOrigin: 'top left',
            position: 'relative',
            zIndex: 10,
          },
          config: animatedSpringConfig,
        };
      }
    });

    instances += 1;
  };

  const closeLayer = () => {
    const bounds = lastKnownTriggerBounds.current;
    // TODO: fix it to use direction context
    const isRTL = document.documentElement.dir === 'rtl';

    layerStylesAPI.start({
      to: {
        insetInlineStart: isRTL ? bounds.left * -1 : bounds.left,
        top: bounds.top,
        transform: `scale(${bounds.width / windowWidth}, ${bounds.height / windowHeight})`,
        pointerEvents: 'none',
        opacity: -1,
        transformOrigin: 'top left',
        zIndex: -1,
      },
      config: animatedSpringConfig,
    });

    triggerStylesAPI.start({
      to: initialTriggerStyles,
      onRest: () => {
        setIsOpen(false);
        exitCallback?.();
      },
      config: animatedSpringConfig,
    });

    instances = Math.max(instances - 1, 0);
  };

  const content =
    typeof layerProps.content === 'function' ? layerProps.content(closeLayer) : layerProps.content;

  const displayContent = layerProps.placeholder == null || isOpen || hasBeenOpened.current;

  return {
    layer: (
      <Layer
        content={displayContent ? content : layerProps.placeholder}
        closeLabel={layerProps.closeLabel}
        title={layerProps.title}
        rightButtonProps={
          layerProps.rightButtonProps && {
            ...layerProps.rightButtonProps,
            onClick: layerProps.rightButtonProps?.onClick || closeLayer,
          }
        }
        leftButtonProps={layerProps.leftButtonProps}
        style={layerStyles}
        onClose={closeLayer}
      />
    ),
    openLayer,
    triggerStyles,
    target: lastKnownRef.current,
  };
};

export default useLayerGenerator;
