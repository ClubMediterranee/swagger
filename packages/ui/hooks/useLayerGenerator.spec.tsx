import { Globals } from '@react-spring/web';
import { render, renderHook, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { act } from 'react-dom/test-utils';

import useLayerGenerator from './useLayerGenerator';

type input = Record<string, unknown> | ((index: number) => Record<string, unknown>);

const layerStylesAPIMock = {
  start: jest.fn((i: input) => {
    if (typeof i === 'function') {
      return i(0);
    }

    return i;
  }),
};
const triggerStylesAPIMock = {
  start: jest.fn((i: input) => {
    if (typeof i === 'function') {
      return i(0);
    }

    return i;
  }),
};

jest.mock('react-dom', () => {
  const original = jest.requireActual('react-dom');
  return {
    ...original,
    createPortal: jest.fn((modal) => modal),
  };
});
jest.mock('@react-spring/web', () => {
  const original = jest.requireActual('@react-spring/web');
  return {
    ...original,
    useScroll: () => ({ scrollY: { to: (cb: (x: number) => number) => cb(5) } }),
    useSprings: () => [[{ pointerEvents: 'auto' }], triggerStylesAPIMock],
    useSpring: () => [{ pointerEvents: 'auto' }, layerStylesAPIMock],
  };
});

describe('useLayerGenerator()', () => {
  beforeAll(() => {
    Globals.assign({
      skipAnimation: true,
    });
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const rect = {
    x: 0,
    y: 0,
    width: 250,
    height: 100,
    top: 20,
    left: 50,
  } as DOMRect;

  const layerProps = {
    content: <div>Layer content!</div>,
    title: 'Layer title!',
    closeLabel: 'Close layer',
  };

  it('returns an object', () => {
    const { result } = renderHook(() => useLayerGenerator(layerProps, 1));

    const expected = {
      layer: expect.anything(),
      openLayer: expect.any(Function),
      target: null,
      triggerStyles: [{ pointerEvents: 'auto' }],
    };

    expect(result.current).toEqual(expected);
    expect(result.current.layer).toMatchSnapshot('Layer Snapshot');
  });

  describe('openLayer()', () => {
    it('call springs functions', () => {
      const { result } = renderHook(() => useLayerGenerator(layerProps, 1));

      act(() => result.current.openLayer(rect, 0));

      expect(layerStylesAPIMock.start).toHaveBeenCalledTimes(1);
      expect(triggerStylesAPIMock.start).toHaveBeenCalledTimes(1);
      expect(layerStylesAPIMock.start).toHaveBeenCalledWith({
        config: { friction: 30, mass: 1.5, tension: 180 },
        from: {
          insetInlineStart: 50,
          top: 20,
          transform: 'scale(0.244140625, 0.13020833333333334)',
          pointerEvents: 'none',
          opacity: 0.2,
          transformOrigin: 'top left',
          zIndex: 1,
          x: '0%',
        },
        to: {
          insetInlineStart: 0,
          top: 0,
          transform: 'scale(1, 1)',
          pointerEvents: 'auto',
          opacity: 1,
          transformOrigin: 'top left',
          zIndex: 1,
        },
      });
      expect(triggerStylesAPIMock.start).toHaveReturnedWith({
        config: { mass: 1.5, tension: 180, friction: 30 },
        from: {
          insetInlineStart: 0,
          top: 0,
          transform: 'scale(1, 1)',
          willChange: 'transform, opacity, inset, filter',
          pointerEvents: 'auto',
          opacity: 1,
          filter: 'blur(0px)',
          transformOrigin: 'top left',
          position: 'relative',
          zIndex: 10,
        },
        to: {
          insetInlineStart: -50,
          top: -20,
          transform: 'scale(4.096, 4.096)',
          pointerEvents: 'none',
          opacity: 0,
          filter: 'blur(10px)',
          transformOrigin: 'top left',
          position: 'relative',
          zIndex: 10,
        },
      });
    });
  });

  describe('closeLayer()', () => {
    it('call springs functions', async () => {
      const { result } = renderHook(() => useLayerGenerator(layerProps, 1));
      render(result.current.layer);

      act(() => result.current.openLayer(rect, 0));

      const closeLayerButton = screen.getByRole('button', { name: 'Close layer' });

      await userEvent.click(closeLayerButton);

      expect(layerStylesAPIMock.start).toHaveBeenCalledTimes(2);
      expect(triggerStylesAPIMock.start).toHaveBeenCalledTimes(2);
      expect(layerStylesAPIMock.start).toHaveBeenNthCalledWith(2, {
        config: { friction: 30, mass: 1.5, tension: 180 },
        to: {
          insetInlineStart: 50,
          top: 20,
          transform: 'scale(0.244140625, 0.13020833333333334)',
          pointerEvents: 'none',
          opacity: -1,
          transformOrigin: 'top left',
          zIndex: -1,
        },
      });
      expect(triggerStylesAPIMock.start).toHaveNthReturnedWith(2, {
        config: { friction: 30, mass: 1.5, tension: 180 },
        onRest: expect.any(Function),
        to: {
          filter: 'blur(0px)',
          insetInlineStart: 0,
          opacity: 1,
          pointerEvents: 'auto',
          position: 'relative',
          top: 0,
          transform: 'scale(1, 1)',
          transformOrigin: 'top left',
          willChange: 'transform, opacity, inset, filter',
          zIndex: 0,
        },
      });
    });
  });
});
