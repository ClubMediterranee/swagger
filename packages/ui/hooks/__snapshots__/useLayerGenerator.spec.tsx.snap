// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useLayerGenerator() returns an object: Layer Snapshot 1`] = `
<Layer
  closeLabel="Close layer"
  content={
    <div>
      Layer content!
    </div>
  }
  onClose={[Function]}
  style={
    {
      "pointerEvents": "auto",
    }
  }
  title="Layer title!"
/>
`;
